tools/clang/unittests/HLSLExec/ExecutionTest.cpp

```diff
@@ -2462,6 +2695,758 @@ TEST_F(ExecutionTest, LifetimeIntrinsicTest) {
+    accelInputs.NumDescs = 1;
+    accelInputs.Flags = buildFlags;
+
+    D3D12_RAYTRACING_ACCELERATION_STRUCTURE_PREBUILD_INFO prebuildInfo = {};
+    pDevice->GetRaytracingAccelerationStructurePrebuildInfo(&accelInputs,
+                                                            &prebuildInfo);
+
+    ReallocScratchResource(pDevice, &scratchResource,
+                           prebuildInfo.ScratchDataSizeInBytes);
+    AllocateBuffer(
+        pDevice, prebuildInfo.ResultDataMaxSizeInBytes, &blasResource, true,
+        D3D12_RESOURCE_STATE_RAYTRACING_ACCELERATION_STRUCTURE, L"BLAS");
+
+    D3D12_BUILD_RAYTRACING_ACCELERATION_STRUCTURE_DESC buildDesc = {};
+    buildDesc.Inputs = accelInputs;
+    buildDesc.ScratchAccelerationStructureData =
+        scratchResource->GetGPUVirtualAddress();
+    buildDesc.DestAccelerationStructureData =
+        blasResource->GetGPUVirtualAddress();
+
+    pCommandList->BuildRaytracingAccelerationStructure(&buildDesc, 0, 0);
+    CD3DX12_RESOURCE_BARRIER barrier =
+        CD3DX12_RESOURCE_BARRIER::UAV(blasResource);
+    pCommandList->ResourceBarrier(1, (const D3D12_RESOURCE_BARRIER *)&barrier);
+  }
+
+  pCommandList->Close();
+  ExecuteCommandList(pCommandQueue, pCommandList);
+  WaitForSignal(pCommandQueue, FO);
+
+  VERIFY_SUCCEEDED(pCommandAllocator->Reset());
+  VERIFY_SUCCEEDED(pCommandList->Reset(pCommandAllocator, nullptr));
+
+  // Build TLAS.
+  {
+    D3D12_RAYTRACING_INSTANCE_DESC instanceDesc = {};
+    instanceDesc.Transform[0][0] = instanceDesc.Transform[1][1] =
+        instanceDesc.Transform[2][2] = 1;
+    instanceDesc.InstanceMask = 1;
+    instanceDesc.AccelerationStructure = blasResource->GetGPUVirtualAddress();
+    AllocateUploadBuffer(pDevice, &instanceDesc, sizeof(instanceDesc),
+                         &instanceDescs, L"instanceDescs");
+
+    D3D12_RAYTRACING_ACCELERATION_STRUCTURE_BUILD_FLAGS buildFlags =
+        D3D12_RAYTRACING_ACCELERATION_STRUCTURE_BUILD_FLAG_PREFER_FAST_BUILD;
+
+    D3D12_BUILD_RAYTRACING_ACCELERATION_STRUCTURE_INPUTS accelInputs = {};
+    accelInputs.Type = D3D12_RAYTRACING_ACCELERATION_STRUCTURE_TYPE_TOP_LEVEL;
+    accelInputs.DescsLayout = D3D12_ELEMENTS_LAYOUT_ARRAY;
+    accelInputs.NumDescs = 1;
+    accelInputs.Flags = buildFlags;
+    accelInputs.InstanceDescs = instanceDescs->GetGPUVirtualAddress();
+
+    D3D12_RAYTRACING_ACCELERATION_STRUCTURE_PREBUILD_INFO prebuildInfo = {};
+    pDevice->GetRaytracingAccelerationStructurePrebuildInfo(&accelInputs,
+                                                            &prebuildInfo);
+
+    AllocateBuffer(
+        pDevice, prebuildInfo.ResultDataMaxSizeInBytes, &tlasResource, true,
+        D3D12_RESOURCE_STATE_RAYTRACING_ACCELERATION_STRUCTURE, L"TLAS");
+
+    D3D12_BUILD_RAYTRACING_ACCELERATION_STRUCTURE_DESC buildDesc = {};
+    buildDesc.Inputs = accelInputs;
+    buildDesc.ScratchAccelerationStructureData =
+        scratchResource->GetGPUVirtualAddress();
```

Comment:
A check is needed to allocate a larger scratch buffer here if the TLAS build needs it. But this can't deallocate an older one since the build hasn't even been kicked off yet, so this probably should just allocate a second TLAS scratch buffer.